src/main/java/mekanism/common/integration/framedblocks/ChemicalCamoContainerFactory.java

```diff
@@ -0,0 +1,143 @@
+        Chemical chemical = camo.getChemical();
+        tag.putInt("chemical", MekanismAPI.CHEMICAL_REGISTRY.getId(chemical));
+    }
+
+    @Override
+    protected ChemicalCamoContainer readFromNetwork(CompoundTag tag) {
+        Chemical chemical = MekanismAPI.CHEMICAL_REGISTRY.byId(tag.getInt("chemical"));
+        return new ChemicalCamoContainer(chemical);
+    }
+
+    @Override
+    @Nullable
+    public ChemicalCamoContainer applyCamo(Level level, BlockPos pos, Player player, ItemStack stack) {
+        IChemicalHandler handler = Capabilities.CHEMICAL.getCapability(stack);
+        if (handler == null || handler.getChemicalTanks() <= 0) {
+             return null;
+        }
+
+        ChemicalStack chemical = handler.getChemicalInTank(0);
+        if (!isValidChemical(chemical.getChemical(), player)) {
+            return null;
+        }
+
+        if (!player.isCreative() && ConfigView.Server.INSTANCE.shouldConsumeCamoItem()) {
+            ChemicalStack extracted = handler.extractChemical(0, FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT, Action.SIMULATE);
+            if (extracted.getAmount() != FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT) {
+                return null;
+            }
+
+            if (!level.isClientSide()) {
+                handler.extractChemical(0, FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT, Action.EXECUTE);
+            }
+        }
+
+        return new ChemicalCamoContainer(chemical.getChemical());
+    }
+
+    @Override
+    public boolean removeCamo(Level level, BlockPos pos, Player player, ItemStack stack, ChemicalCamoContainer camo) {
+        if (stack.isEmpty()) {
+            return false;
+        }
+
+        IChemicalHandler handler = Capabilities.CHEMICAL.getCapability(stack);
+        if (handler == null || handler.getChemicalTanks() <= 0) {
+            return false;
+        }
+
+        ChemicalStack chemical = camo.getChemical().getStack(FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT);
+        if (handler.insertChemical(chemical, Action.SIMULATE).isEmpty()) {
+            if (!level.isClientSide() && !player.isCreative() && ConfigView.Server.INSTANCE.shouldConsumeCamoItem()) {
+                handler.insertChemical(chemical, Action.EXECUTE);
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean canTriviallyConvertToItemStack() {
+        return false;
+    }
+
+    @Override
+    public ItemStack dropCamo(ChemicalCamoContainer camo) {
+        return ItemStack.EMPTY;
+    }
+
+    @Override
+    public boolean validateCamo(ChemicalCamoContainer camo) {
+        return isValidChemical(camo.getChemical(), null);
+    }
+
+    private static boolean isValidChemical(Chemical chemical, @Nullable Player player) {
+        if (chemical.isEmptyType()) {
+            return false;
+        }
+        if (chemical.isRadioactive()) {
```

Comment:
We probably want to use `Chemical#hasAttributesWithValidation` instead (with a slightly different validation message). That way if we later add any attributes other than radiation that need validation or an addon does, we can respect that here.

Reply:
That's fair, do you have an idea for a sensible, more generic message?

Reply:
Chemicals that require special handling cannot be put in a framed block

Or something along those lines